### PR TITLE
Document the Value Propagation optimization

### PR DESCRIPTION
**Pull request by fjeremic**
Use the transcription and outline in #4927 to fully document how the
Value Propagation optimization works at a high level, giving details
and references to the code where needed.

Special thanks to @vijaysun-omr for describing the optimization is such
great detail which enabled this document to be written.

[skip ci]

Closes: #4927

Signed-off-by: Filip Jeremic <fjeremic@ca.ibm.com>

### CODE REVIEW

**Comment by fjeremic** [comment]
Note to reviewers the document is much more pleasant to read and review using the "rich diff" display:
https://github.com/eclipse/omr/pull/5101/files?short_path=08e068b#diff-08e068b7f0ec0f0b9bb7638f050b0995

File: doc/compiler/optimizer/ValuePropagation.md [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+This is a very high level summary of what value numbering is. There are excellent resources in literature or online to learn more about value numbering.
+
+## Constraints and Handlers
+
+[Value propagation constraints](https://github.com/eclipse/omr/blob/25f479d2b671c2a0a03156cc0a48024a46d2e3c2/compiler/optimizer/VPConstraint.hpp) are the fundamental classes which are flowed around by the value propagation analysis. Their purpose is the represent constraints determined by the analysis on _value numbers_. If for example we had an expression such as `x = 5` then we would create a constraint which represents the constant value of `5` and we would associate this constraint with the value numbers for the nodes representing the left hand side (`iconst`) and right hand side (`istore`) nodes. Remember that these two nodes would have the same value number because we can prove that they are equal. The VP handlers consume these relations between value numbers and constraints to optimize the program.
```
Comment (vijaysun-omr) [inline doc/compiler/optimizer/ValuePropagation.md:37]: `left hand side (`iconst`) and right hand side (`istore`) nodes` should probably be the other way around

File: doc/compiler/optimizer/ValuePropagation.md [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+[Value propagation handlers](https://github.com/eclipse/omr/blob/25f479d2b671c2a0a03156cc0a48024a46d2e3c2/compiler/optimizer/VPHandlers.cpp) are functions which operate on the set of constraints associated with the nodes children and produce constraints for the parent node. There are many handler functions associated with many opcodes. The [value propagation table](https://github.com/eclipse/omr/blob/ad822fbd7f17f1aa1fddafa355cb18fac4247a46/compiler/optimizer/ValuePropagationTable.hpp) describes which function to execute for each particular opcode. Some opcodes are not interesting and all they do is constrain their children, while others perform an interesting operation.
+
```
Comment (vijaysun-omr) [inline doc/compiler/optimizer/ValuePropagation.md:40]: `associated with the nodes children` might be missing an apostrophe.

File: doc/compiler/optimizer/ValuePropagation.md (hunk 1 above)
Comment (fjeremic) [reply doc/compiler/optimizer/ValuePropagation.md]: Fixed in [force-push](https://github.com/eclipse/omr/compare/76b1afbc94d81d8c76ee613e2e89e6c3aa4aab43..9fb1b0b92e8935d254254edca4bd50d3a6c93e48).

File: doc/compiler/optimizer/ValuePropagation.md (hunk 2 above)
Comment (fjeremic) [reply doc/compiler/optimizer/ValuePropagation.md:40]: Fixed in [force-push](https://github.com/eclipse/omr/compare/9fb1b0b92e8935d254254edca4bd50d3a6c93e48..d9691be5fa5c4c985b7f2c6c4493e532b222cb40).

File: doc/compiler/optimizer/ValuePropagation.md [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+A `merge` of two constraints will weaken the constraint. For example consider the following program:
+
+```
+if (...) {
+    x = 10;
+} else {
+    x = 5;
+}
+
```
[the remark this answers is not in the document]
Comment (fjeremic) [reply doc/compiler/optimizer/ValuePropagation.md:75]: Fixed in [force-push](https://github.com/eclipse/omr/compare/d9691be5fa5c4c985b7f2c6c4493e532b222cb40..5c1aa757ca3e116307e60404b9d3b0743e2ce03f).

File: doc/compiler/optimizer/ValuePropagation.md [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+### Type Constraints
+
+For typed languages it is often beneficial to flow type constraints. Knowing type information about a particular value number at a given point in the program can allow us to devirtualize calls or eliminate type check and exception edges.
+
+There are three different constraint classes which flow this information in VP, they are the [fixed class](https://github.com/eclipse/omr/blob/25f479d2b671c2a0a03156cc0a48024a46d2e3c2/compiler/optimizer/VPConstraint.hpp#L655-L674) constraint, [resolved class](https://github.com/eclipse/omr/blob/25f479d2b671c2a0a03156cc0a48024a46d2e3c2/compiler/optimizer/VPConstraint.hpp#L628-L653) constraint, and [unresolved class](https://github.com/eclipse/omr/blob/25f479d2b671c2a0a03156cc0a48024a46d2e3c2/compiler/optimizer/VPConstraint.hpp#L706-L736) constraints.
+
+Fixed class constraints simply express that a value number has a specific, fixed, type `T`. We can know this information perhaps because we have seen the allocation of the value number expression or perhaps we have seen an explicit cast into type `T`. Knowing the exact type of a value number expression is the most useful.
+
```
Comment (vijaysun-omr) [inline doc/compiler/optimizer/ValuePropagation.md:98]: An explicit cast into type `T` is not usually enough to generate a fixed class constraint. The type that we cast to must be `final` in order for the constraint to be a fixed one, otherwise it just stays a resolved class constraint.

File: doc/compiler/optimizer/ValuePropagation.md [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+Resolved class constraints express that a value number is of type `T` or any subclass of `T`. These constraints are useful for eliminating additional type checks on the value number at some later point in the program.
+
+Unresolved class constraints are least interesting and express that the class type of the value number has not been loaded by the runtime and is unknown. They simply carry around the string representing the class of the object.
+
+### Null and NonNull Constraints
+
+There are two notions flowed through VP for reference types in regards to nullness. The [null](https://github.com/eclipse/omr/blob/25f479d2b671c2a0a03156cc0a48024a46d2e3c2/compiler/optimizer/VPConstraint.hpp#L746-L760) constraint, and the [non-null](https://github.com/eclipse/omr/blob/25f479d2b671c2a0a03156cc0a48024a46d2e3c2/compiler/optimizer/VPConstraint.hpp#L762-L775) constraint. These constraints are useful for tracking value numbers which are certainly known to be either `null` or certainly not `null`. Knowing such information can eliminate implicit and explicit null checks or open doors for further simplifications.
+
```
Comment (vijaysun-omr) [inline doc/compiler/optimizer/ValuePropagation.md:106]: These also help fold away comparisons (ifacmpeq/ne) that test against null.

File: doc/compiler/optimizer/ValuePropagation.md [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+### Absolute vs. Relative Constraints
+
+Absolute constraints are explicit constraints about a particular value number. For example a type constraint, a range constraint, a null constraint are all examples or absolute constraints.
+
+Relative constraints are constraints that connect one value number to another value number. For example if we were to encounter a guarded comparison between `x` and `y`, say `x < y` then we would create a relative constraint between the value numbers of `x` and `y` to specify that on that path the value number for `x` is less than the value number for `y`.
+
+Relative constraints are useful when neither `x` nor `y` are known to be constant, and we still want to propagate information about the relation of their value numbers. This is useful if for example we encounter another condition which is a successor of the previous check which asked if `x < y + 5`. We want to be able to fold away such expressions even though we don't know any absolute constraints about the value numbers of `x` nor `y`.
+
```
Comment (vijaysun-omr) [inline doc/compiler/optimizer/ValuePropagation.md:116]: Sometimes overflow possibilities can mean that we cannot fold away compares such as the above but the general idea is conveyed by this example.

File: doc/compiler/optimizer/ValuePropagation.md [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+## Local vs. Global Value Propagation
+
+There are two modes in which the value propagation optimization can run; local and global. Both modes run using the same constraints and VP handlers. What is different is the scope at which they operate.
+
+Local value propagation processes the program one block at a time. If block extension has ran it processes the program one extended block at a time. Information is not propagated between blocks, so any information we learned is discarded after arriving at the end of the (extended) basic block. This of course makes the local value propagation optimization restrictive in what it can accomplish.
+
```
Comment (vijaysun-omr) [inline doc/compiler/optimizer/ValuePropagation.md:122]: "restrictive" but also "cheaper" from a compile time perspective.

Comment (vijaysun-omr) [reply doc/compiler/optimizer/ValuePropagation.md:122]: edit : you make this point a bit later in the description

File: doc/compiler/optimizer/ValuePropagation.md [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+Because local value propagation runs at this reduced scope there is no need to calculate value numbers for all the nodes in the program. We can simply use the global index of the node as the value number, since if local value propagation runs after local common subexpression elimination (CSE), the nodes within a block will have already been commened up and they will have the same global index. Retrieving the global index is very cheap, and in addition other expensive metadata such as use-def info need not be computed to run local value propagation.
+
+This makes it the perfect candidate to run frequently and at low optimization levels.
+
+Global value propagation on the other hand is a global optimization which requires both global value numbering and use-def information to be computed. In the context of the OMR compiler global means at the scope of the current compilation unit, not intra-procedure. The extra metadata information which needs to be computed is relatively expensive and this is why the global value propagation optimization is only run at higher optimization levels (warm and above). 
+
```
Comment (vijaysun-omr) [inline doc/compiler/optimizer/ValuePropagation.md:128]: I think you mean "inter procedural" when you said "intra-procedure"

File: doc/compiler/optimizer/ValuePropagation.md [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+if (...) {
+    ...
+} else {
+    ...
+}
+```
+
+The value number of the right hand side of the statement `i = 5` has a constraint which will cannot be changed. So why bother to propagate such a constraint through the successor edges of the subsequent `if` statement. Rather than doing this we keep track of a global constraint list to which we add such global constraints which are valid throughout the entire analysis. This saves on memory footprint and time to copy the constraints around as the analysis flows through the method.
+
```
Comment (vijaysun-omr) [inline doc/compiler/optimizer/ValuePropagation.md:175]: "will cannot be changed" is a typo

File: doc/compiler/optimizer/ValuePropagation.md [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+These check nodes can be quite common in languages such as Java where we would have nodes representing null checks (`NULLCHK`), bound checks (`BNDCHK`), etc. All of these check nodes can cause an exception to occur (`NullPointerException` and `ArrayIndexOutOfBoundsException` respectively). In the compiler we create exception edges from blocks containing such excepting nodes to their corresponding catch blocks.
+
+Optimizations such as value propagation which walk the control flow graph need to deal with such exception edges. The reason for this is because execution can resume if the catch block handles the exception and falls through to the next block. We need to properly keep track of the constraints at the merge point of the catch block and the fall-through of the try block.
+
+To do this the value propagation optimization merges the set of constraints known at the time we are examining the check node onto the exception edge. Each new check node we encounter we perform another merge operation because the set of constraints may have changed between two check nodes which could be several tree tops apart in the _same_ block.
```
Comment (vijaysun-omr) [inline doc/compiler/optimizer/ValuePropagation.md:195]: This is accurate but I did want to state that there can be nodes other than check nodes that can cause exceptions. One case in point is call nodes. So this description may be better talking about nodes that "may cause an exception" rather than "check" nodes.

**Comment by vijaysun-omr** [comment]
Looks very good @fjeremic and I thank you again for not only taking the time to document this, but doing so very accurately and correctly. I am sure it will be of use to the OMR community down the line.

File: doc/compiler/optimizer/ValuePropagation.md (hunk 10 above)
Comment (fjeremic) [reply doc/compiler/optimizer/ValuePropagation.md]: Fixed in [force-push](https://github.com/eclipse/omr/compare/5c1aa757ca3e116307e60404b9d3b0743e2ce03f..127bdeabc6860a9f211446b05d6210a084512605).

File: doc/compiler/optimizer/ValuePropagation.md (hunk 9 above)
Comment (fjeremic) [reply doc/compiler/optimizer/ValuePropagation.md:175]: Fixed in [force-push](https://github.com/eclipse/omr/compare/5c1aa757ca3e116307e60404b9d3b0743e2ce03f..127bdeabc6860a9f211446b05d6210a084512605).

File: doc/compiler/optimizer/ValuePropagation.md (hunk 8 above)
Comment (fjeremic) [reply doc/compiler/optimizer/ValuePropagation.md:128]: Fixed in [force-push](https://github.com/eclipse/omr/compare/5c1aa757ca3e116307e60404b9d3b0743e2ce03f..127bdeabc6860a9f211446b05d6210a084512605).

File: doc/compiler/optimizer/ValuePropagation.md (hunk 7 above)
Comment (fjeremic) [reply doc/compiler/optimizer/ValuePropagation.md:122]: Fixed in [force-push](https://github.com/eclipse/omr/compare/5c1aa757ca3e116307e60404b9d3b0743e2ce03f..127bdeabc6860a9f211446b05d6210a084512605). I added it just to be extra explicit that the cost is from both memory footprint and compile time.

File: doc/compiler/optimizer/ValuePropagation.md (hunk 6 above)
Comment (fjeremic) [reply doc/compiler/optimizer/ValuePropagation.md:116]: Fixed in [force-push](https://github.com/eclipse/omr/compare/5c1aa757ca3e116307e60404b9d3b0743e2ce03f..127bdeabc6860a9f211446b05d6210a084512605).

File: doc/compiler/optimizer/ValuePropagation.md (hunk 5 above)
Comment (fjeremic) [reply doc/compiler/optimizer/ValuePropagation.md:106]: Fixed in [force-push](https://github.com/eclipse/omr/compare/5c1aa757ca3e116307e60404b9d3b0743e2ce03f..127bdeabc6860a9f211446b05d6210a084512605).

File: doc/compiler/optimizer/ValuePropagation.md (hunk 4 above)
Comment (fjeremic) [reply doc/compiler/optimizer/ValuePropagation.md:98]: Fixed in [force-push](https://github.com/eclipse/omr/compare/5c1aa757ca3e116307e60404b9d3b0743e2ce03f..127bdeabc6860a9f211446b05d6210a084512605).

**Comment by vijaysun-omr** [comment]
Checks have passed and my requested changes are all done. Merging.
